src/tokenizer/shell.cpp

```diff
@@ -0,0 +1,526 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2021 Datadog, Inc.
+
+#include "tokenizer/shell.hpp"
+#include "utils.hpp"
+
+using namespace std::literals;
+
+namespace ddwaf {
+
+namespace {
+re2::RE2 redirection_regex(
+    R"((&>>?|(?:[0-9]?>(?:\||>|&[0-9]?-?)?)|(?:[0-9]?<(?:<(?:-|<)?|&[0-9]?-?|>)?)))");
+
+bool is_var_char(char c) { return ddwaf::isalnum(c) || c == '_'; }
+
+bool is_field_char(char c)
+{
+    static constexpr std::string_view known_chars = "`${}[]()'\"#=|<>;&\n\t ";
+    return known_chars.find(c) == std::string_view::npos;
+}
+
+bool is_space_char(char c) { return c == ' ' || c == '\t'; }
+
+std::vector<shell_token> find_executables_and_strip_whitespaces(std::vector<shell_token> &tokens)
+{
+    // The scope within the command, this helps identify high level constructs
+    // which end up evaluating as part of a command, e.g. an executable
+    // generated from a command substitution
+    enum class command_scope {
+        variable_definition_or_executable,
+        variable_definition,
+        arguments,
+        none,
+    };
+
+    std::size_t read = 0;
+    std::size_t write = 0;
+
+    std::vector<command_scope> command_scope_stack{
+        command_scope::variable_definition_or_executable};
+    for (; read < tokens.size(); read++) {
+        auto &token = tokens[read];
+        auto &scope = command_scope_stack.back();
+
+        switch (token.type) {
+        case shell_token_type::backtick_substitution_open:
+        case shell_token_type::command_substitution_open:
+        case shell_token_type::process_substitution_open:
+        case shell_token_type::compound_command_open:
+        case shell_token_type::subshell_open:
+            if (scope == command_scope::variable_definition_or_executable) {
+                // These tokens have nested commands which will be interpreted
+                // as the executable, so we can consider the executable of this
+                // scope "found", albeit detections will rely on nested
+                // executables, rather than considering this the executable.
```

Comment:
This refers to the fact that if a command consists of `$(echo ls) -l`, the command substitution `$()` is the actual executable, however instead of tagging the whole thing as the executable, we assume that we're already tagging `echo` as the executable, so tagging the whole thing might not be necessary.

I'll clarify the comment.